lib/security/JweToken.ts

```diff
@@ -105,18 +246,35 @@ export default class JweToken extends JoseToken {
    */
   public async decrypt (jwk: PrivateKey): Promise<string> {
     // following steps for JWE Decryption in RFC7516 section 5.2
-    // 1. Parse JWE for components: BASE64URL(UTF8(JWE Header)) || '.' || BASE64URL(JWE Encrypted Key) || '.' ||
-    //    BASE64URL(JWE Initialization Vector) || '.' || BASE64URL(JWE Ciphertext) || '.' ||
-    //    BASE64URL(JWE Authentication Tag)
-    const base64EncodedValues = this.content.split('.');
-    // 2. Base64url decode the encoded header, encryption key, iv, ciphertext, and auth tag
-    const headerString = Base64Url.decode(base64EncodedValues[0]);
-    const encryptedKey = Buffer.from(Base64Url.toBase64(base64EncodedValues[1]), 'base64');
-    const iv = Buffer.from(Base64Url.toBase64(base64EncodedValues[2]), 'base64');
-    const ciphertext = Buffer.from(Base64Url.toBase64(base64EncodedValues[3]), 'base64');
-    const authTag = Buffer.from(Base64Url.toBase64(base64EncodedValues[4]), 'base64');
-    // 3. let the JWE Header be a JSON object
-    const headers = JSON.parse(headerString);
+    let encryptedKey: Buffer;
+    let iv: Buffer;
+    let ciphertext: Buffer;
+    let authTag: Buffer;
+    let aad: string;
+
+    if (this.isFlattenedJSONSerialized) {
+      // use the pre-parsed contents.
+      const headerString = this.protected || '';
+      encryptedKey = Buffer.from(Base64Url.toBase64(this.encrypted_key!), 'base64');
+      iv = Buffer.from(Base64Url.toBase64(this.iv!), 'base64');
+      ciphertext = Buffer.from(Base64Url.toBase64(this.content), 'base64');
+      authTag = Buffer.from(Base64Url.toBase64(this.tag!), 'base64');
+      aad = this.aad ? headerString + '.' + this.aad : headerString;
+    } else {
+      // 1. Parse JWE for components: BASE64URL(UTF8(JWE Header)) || '.' || BASE64URL(JWE Encrypted Key) || '.' ||
+      //    BASE64URL(JWE Initialization Vector) || '.' || BASE64URL(JWE Ciphertext) || '.' ||
+      //    BASE64URL(JWE Authentication Tag)
+      const base64EncodedValues = this.content.split('.');
+
+      // 2. Base64url decode the encoded header, encryption key, iv, ciphertext, and auth tag
+      encryptedKey = Buffer.from(Base64Url.toBase64(base64EncodedValues[1]), 'base64');
+      iv = Buffer.from(Base64Url.toBase64(base64EncodedValues[2]), 'base64');
+      ciphertext = Buffer.from(Base64Url.toBase64(base64EncodedValues[3]), 'base64');
+      authTag = Buffer.from(Base64Url.toBase64(base64EncodedValues[4]), 'base64');
```

Comment:
Given the frequencey of this call pattern, it's probably a good time to add a method for this in a utilities class. (Perhaps in Base64Url class if this was actually supposed to be for Base64URL encoding and not Base64).